flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/MultipleInputStreamTaskChainedSourcesCheckpointingTest.java

```diff
@@ -330,6 +338,64 @@ public void testRpcTriggerCheckpointWithSourceChain() throws Exception {
         }
     }
 
+    @Test
+    public void testSkipExecutionsIfFinishedOnRestoreWithSourceChained() throws Exception {
+        OperatorID firstSourceOperatorId = new OperatorID();
+        OperatorID secondSourceOperatorId = new OperatorID();
+        OperatorID nonSourceOperatorId = new OperatorID();
+
+        try (StreamTaskMailboxTestHarness<String> testHarness =
+                new StreamTaskMailboxTestHarnessBuilder<>(
+                                MultipleInputStreamTask::new, BasicTypeInfo.STRING_TYPE_INFO)
+                        .modifyStreamConfig(config -> config.setCheckpointingEnabled(true))
+                        .modifyExecutionConfig(ExecutionConfig::enableObjectReuse)
+                        .addInput(BasicTypeInfo.INT_TYPE_INFO)
+                        .addSourceInput(
+                                firstSourceOperatorId,
+                                new SourceOperatorFactory<>(
+                                        new SourceOperatorStreamTaskTest.LifeCycleMonitorSource(
+                                                Boundedness.BOUNDED, 1),
```

Comment:
nit: as a safety precaution let's use the `CONTINUOUS_UNBOUNDED` here, theoretically for bounded sources we might not have checkpointing

I know at this point we will, but let's do it for my peaces of mind ;)